runtime/src/iree/hal/drivers/cuda/dynamic_symbols.c

```diff
@@ -65,10 +100,45 @@ iree_status_t iree_hal_cuda_dynamic_symbols_initialize(
   return status;
 }
 
+iree_status_t iree_hal_cuda_nccl_dynamic_symbols_initialize(
+    iree_allocator_t host_allocator,
+    iree_hal_cuda_dynamic_symbols_t* out_syms) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  if (!out_syms->cuda_library) {
+    iree_status_t status =
+        iree_make_status(IREE_STATUS_FAILED_PRECONDITION,
+                         "CUDA dynamic symbols are not loaded first.");
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, status, "iree_hal_cuda_nccl_dynamic_symbols_initialize");
+  }
+
+  out_syms->nccl_library = NULL;
+  iree_status_t status = iree_ok_status();
+  status = iree_dynamic_library_load_from_files(
+      IREE_ARRAYSIZE(kNCCLLoaderSearchNames), kNCCLLoaderSearchNames,
+      IREE_DYNAMIC_LIBRARY_FLAG_NONE, host_allocator, &out_syms->nccl_library);
+  if (iree_status_is_not_found(status)) {
+    iree_status_ignore(status);
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_make_status(IREE_STATUS_UNAVAILABLE,
+                             "NCCL runtime library not available; ensure "
+                             "installed and on path"));
+  }
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_cuda_nccl_dynamic_symbols_resolve_all(out_syms);
+  } else {
+    iree_hal_cuda_dynamic_symbols_deinitialize(out_syms);
```

Comment:
you'll want this to be split so that you still deinitialize if resolving fails:
```c
  if (iree_status_is_ok(status)) {
    status = iree_hal_cuda_nccl_dynamic_symbols_resolve_all(out_syms);
  }
  if (!iree_status_is_ok(status)) {
    iree_hal_cuda_dynamic_symbols_deinitialize(out_syms);
  }
```